testutil/validatormock/meta.go

```diff
@@ -0,0 +1,135 @@
+// Copyright © 2022-2023 Obol Labs Inc. Licensed under the terms of a Business Source License 1.1
+
+package validatormock
+
+import "time"
+
+// sepcMeta defines the spec constants.
+type specMeta struct {
+	GenesisTime   time.Time
+	SlotDuration  time.Duration
+	SlotsPerEpoch uint64
+}
+
+func (m specMeta) SlotStartTime(slot uint64) time.Time {
+	return m.GenesisTime.Add(time.Duration(slot) * m.SlotDuration)
+}
+
+func (m specMeta) EpochFromSlot(slot uint64) metaEpoch {
+	epoch := slot / m.SlotsPerEpoch
+	return metaEpoch{
+		Epoch: epoch,
+		meta:  m,
+	}
+}
+
+func (m specMeta) FirstSlotInEpoch(epoch uint64) metaSlot {
+	return metaSlot{
+		Slot: epoch * m.SlotsPerEpoch,
+		meta: m,
+	}
+}
+
+func (m specMeta) LastSlotInEpoch(epoch uint64) metaSlot {
+	return metaSlot{
+		Slot: (epoch+1)*m.SlotsPerEpoch - 1,
+		meta: m,
+	}
+}
+
+// metaSlot defines a slot with knowledge of spec metadata.
+type metaSlot struct {
+	Slot uint64
+	meta specMeta
+}
+
+func (s metaSlot) StartTime() time.Time {
+	return s.meta.SlotStartTime(s.Slot)
+}
+
+func (s metaSlot) Duration() time.Duration {
+	return s.meta.SlotDuration
+}
+
+func (s metaSlot) Epoch() metaEpoch {
+	return s.meta.EpochFromSlot(s.Slot)
+}
+
+func (s metaSlot) Next() metaSlot {
+	return metaSlot{
+		Slot: s.Slot + 1,
+		meta: s.meta,
+	}
+}
+
+func (s metaSlot) InSlot(t time.Time) bool {
+	startTime := s.StartTime()      // Including
+	endTime := s.Next().StartTime() // Excluding
+
+	return (t.After(startTime) && t.Before(endTime)) || t.Equal(startTime)
+}
+
+func (s metaSlot) FirstInEpoch() bool {
+	return s.Slot == s.Epoch().FirstSlot().Slot
+}
+
+// metaEpoch defines an epoch with knowledge of spec metadata.
+type metaEpoch struct {
```

Comment:
Please write godoc for this type's methods.

Reply:
Done!